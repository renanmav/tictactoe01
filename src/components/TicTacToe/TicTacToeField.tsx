import React from 'react';
import { Image, TouchableWithoutFeedback, View } from 'react-native';
import { t } from 'react-native-tailwindcss';

const s = {
  wrapper: [t.bgTransparent, t.w4_12, t.p1, { aspectRatio: 1 }],
  field: [t.bgBlue900, t.rounded, t.flex1, t.itemsCenter, t.justifyCenter, t.overflowHidden],
  symbol: [t.w8_12, t.objectContain],
  success: [t.bgGreen900],
};

interface TicTacToeFieldProps {
  symbol: 'X' | 'O' | ' ';
  onPress: () => void;
  success?: boolean;
}

const TicTacToeField: React.FC<TicTacToeFieldProps> = ({ symbol, onPress, success = false }) => {
  function renderSymbol() {
    switch (symbol) {
      case 'X':
        return <Image source={require('../../../assets/x.png')} style={s.symbol} testID="X" />;
      case 'O':
        return <Image source={require('../../../assets/o.png')} style={s.symbol} testID="O" />;
      default:
        return null;
    }
  }

  return (
    <View style={s.wrapper}>
      <TouchableWithoutFeedback onPress={onPress} testID="field-touchable">
        <View style={[s.field, success && s.success]} testID="field-background">
          {renderSymbol()}
        </View>
      </TouchableWithoutFeedback>
    </View>
  );
};

export default TicTacToeField;
