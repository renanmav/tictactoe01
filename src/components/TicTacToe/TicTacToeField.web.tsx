import React from 'react';
import { Dimensions, Image, TouchableWithoutFeedback, View } from 'react-native';
import { t } from 'react-native-tailwindcss';

const { width, height } = Dimensions.get('window');

const biggestMeasure = width > height ? width : height;

const s = {
  wrapper: [
    t.bgTransparent,
    { width: biggestMeasure / 3 / 3 - 5, height: biggestMeasure / 3 / 3 - 5, padding: 5 },
  ],
  field: [t.bgBlue900, t.rounded, t.flex1, t.itemsCenter, t.justifyCenter, t.overflowHidden],
  symbol: [t.objectContain, { width: biggestMeasure / 16, height: biggestMeasure / 16 }],
  success: [t.bgGreen900],
};

interface TicTacToeFieldProps {
  symbol: 'X' | 'O' | ' ';
  onPress: () => void;
  success?: boolean;
}

const TicTacToeField: React.FC<TicTacToeFieldProps> = ({ symbol, onPress, success = false }) => {
  function renderSymbol() {
    switch (symbol) {
      case 'X':
        return <Image source={require('../../../assets/x.png')} style={s.symbol} testID="X" />;
      case 'O':
        return <Image source={require('../../../assets/o.png')} style={s.symbol} testID="O" />;
      default:
        return null;
    }
  }

  return (
    <View style={s.wrapper}>
      <TouchableWithoutFeedback onPress={onPress} testID="field-touchable">
        <View style={[s.field, success && s.success]} testID="field-background">
          {renderSymbol()}
        </View>
      </TouchableWithoutFeedback>
    </View>
  );
};

export default TicTacToeField;
