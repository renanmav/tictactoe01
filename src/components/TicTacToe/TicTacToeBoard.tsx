import React from 'react';
import { View } from 'react-native';
import { t } from 'react-native-tailwindcss';

import TicTacToeField from './TicTacToeField';

const s = {
  board: [t.bgBlue800, t.wFull, t.rounded, t.flexRow, t.flexWrap, t.p1, { aspectRatio: 1 }],
};

interface TicTacToeBoardProps {
  board: string;
  onPress(index: number): void;
}

const TicTacToeBoard: React.FC<TicTacToeBoardProps> = ({ board, onPress }) => {
  if (board.length !== 9) {
    throw new Error('Board should have 9 characters');
  }

  return (
    <View style={s.board}>
      {board.split('').map((symbol, index) => (
        <TicTacToeField
          key={index}
          symbol={symbol as any}
          onPress={() => {
            onPress(index);
          }}
        />
      ))}
    </View>
  );
};

export default TicTacToeBoard;
