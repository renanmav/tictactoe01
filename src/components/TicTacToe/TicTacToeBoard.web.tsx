import React from 'react';
import { Dimensions, View } from 'react-native';
import { t } from 'react-native-tailwindcss';

import TicTacToeField from './TicTacToeField';

const { width, height } = Dimensions.get('window');

const biggestMeasure = width > height ? width : height;

const s = {
  board: [
    t.bgBlue800,
    t.rounded,
    t.flexRow,
    t.flexWrap,
    t.justifyCenter,
    t.alignCenter,
    { width: biggestMeasure / 3, height: biggestMeasure / 3, padding: 5 },
  ],
};

interface TicTacToeBoardProps {
  board: string;
  onPress(index: number): void;
  match: number[];
}

const TicTacToeBoard: React.FC<TicTacToeBoardProps> = ({ board, onPress, match }) => {
  if (board.length !== 9) {
    throw new Error('Board should have 9 characters');
  }

  return (
    <View style={s.board} testID="tictactoe-board">
      {board.split('').map((symbol, index) => (
        <TicTacToeField
          key={index}
          symbol={symbol as any}
          onPress={() => {
            onPress(index);
          }}
          success={!!match.includes(index)}
        />
      ))}
    </View>
  );
};

export default TicTacToeBoard;
