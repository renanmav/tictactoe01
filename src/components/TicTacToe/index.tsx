import React from 'react';

import { useBoard } from '../../hooks/useBoard';
import { Difficulties } from '../../screens/SelectDifficulty';

import TicTacToeBoard from './TicTacToeBoard';

interface TicTacToeProps {
  difficulty: Difficulties;
}

const TicTacToe: React.FC<TicTacToeProps> = ({ difficulty }) => {
  const { board, updateBoard, match } = useBoard(difficulty);

  return <TicTacToeBoard board={board} onPress={(i) => updateBoard(i, 'X')} match={match} />;
};

export default TicTacToe;
