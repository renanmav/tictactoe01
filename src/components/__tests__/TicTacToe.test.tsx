import React from 'react';
import { render, fireEvent, wait } from '@testing-library/react-native';
import { FetchMock } from 'jest-fetch-mock';

import TicTacToe from '../TicTacToe';

it('should consider the difficulty prop correctly', async () => {
  (fetch as FetchMock).mockResponseOnce(JSON.stringify({ board: 'XO       ' }));

  const { queryAllByTestId } = render(<TicTacToe difficulty="easy" />);

  const firstField = queryAllByTestId('field-touchable')[0];
  fireEvent.press(firstField);

  await wait(() => {
    expect(fetch).toBeCalledTimes(1);
    expect((fetch as FetchMock).mock.calls).toMatchSnapshot();
  });
});
