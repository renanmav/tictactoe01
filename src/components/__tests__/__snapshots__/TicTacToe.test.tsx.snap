// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should consider the difficulty prop correctly 1`] = `
Array [
  Array [
    "https://tictactoe.api.01card.com.br",
    Object {
      "body": "{\\"board\\":\\"X        \\",\\"level\\":\\"easy\\"}",
      "headers": Object {
        "authorization": "Bearer ck9tzriy6000201l19kdy5d4i",
      },
      "method": "POST",
    },
  ],
]
`;
